tests/test_detectAndMeasure.py

```diff
@@ -189,6 +190,52 @@ def test_measurements_finite(self):
         self._check_values(output.diaSources.getY(), minValue=0, maxValue=ySize)
         self._check_values(output.diaSources.getPsfInstFlux())
 
+    def test_remove_unphysical(self):
+        """
+        """
+
+        # Set up the simulated images
+        noiseLevel = 1.
+        staticSeed = 1
+        xSize = 256
+        ySize = 256
+        kwargs = {"psfSize": 2.4, "xSize": xSize, "ySize": ySize}
+        science, sources = makeTestImage(seed=staticSeed, noiseLevel=noiseLevel, noiseSeed=6,
+                                         nSrc=1, **kwargs)
+        matchedTemplate, _ = makeTestImage(seed=staticSeed, noiseLevel=noiseLevel/4, noiseSeed=7,
+                                           nSrc=1, **kwargs)
+        difference = science.clone()
+        bbox = difference.getBBox()
+        difference.maskedImage -= matchedTemplate.maskedImage
+
+        # Configure the detection Task, and do not remove unphysical sources
+        detectionTask = self._setup_detection(doForcedMeasurement=False, doSkySources=True, nSkySources=20,
+                                              badSourceFlags=["base_PixelFlags_flag_offimage", ])
+
+        # Run detection and check the results
+        diaSources = detectionTask.run(science, matchedTemplate, difference).diaSources
+        badDiaSrc0 = ~bbox.contains(diaSources.getX(), diaSources.getY())
+        nBad0 = np.count_nonzero(badDiaSrc0)
+        # Verify that all sources are physical
+        self.assertEqual(nBad0, 0)
+        # Set a few centroids outside the image bounding box
+        nSetBad = 5
+        for src in diaSources[0: nSetBad]:
+            src["slot_Centroid_x"] += xSize
+            src["slot_Centroid_y"] += ySize
+            src["base_PixelFlags_flag_offimage"] = True
+        # Verify that these sources are outside the image
+        badDiaSrc1 = ~bbox.contains(diaSources.getX(), diaSources.getY())
+        nBad1 = np.count_nonzero(badDiaSrc1)
+        self.assertEqual(nBad1, nSetBad)
+        diaSources2 = detectionTask.removeBadSources(diaSources)
```

Comment:
This test wouldn't have caught the bug I described before (local scoping of the catalog), since it's not looking at the output of `run`. What if instead you set all of `difference.mask` to `BAD`, add the `badCenterAll` flag to the bad flags list, and then just call `detectionTask.run` and confirm that the source is not included in the output? It's crude, but your other tests should be checking that "normal" sources behave correctly, and we don't really care _why_ a given bad flag is set, just that it is.

Reply:
I disagree. If this test were implemented as `detectionTask.removeBadSources(diaSources)` then it would fail because the flagged sources are not removed from `diaSources` in this scope, as you pointed out. Now that the API is for `.removeBadSources` to return a source catalog, it seems appropriate to use that here.

Reply:
If the goal is to test `removeBadSources` by itself as a unittest, then you don't need any of the other infrastructure here. All you need is to create a trivial SourceCatalog with a couple flags (`test_flag_1`, etc.), call `removeBadSources` on that and check that the things you didn't want aren't there. There's no need to create an image or run detection or anything.

If the goal is to test that `run` behaves correctly, then you need to actually test `run` itself. This test as written would not catch `run` not assigning the output of `removeBadSources` to something that gets returned. If this is the goal of this test you don't need to run detectAndMeasure in a way that doesn't remove anything: you're already testing that with the other tests in this class.

Reply:
The code comment here was out of date. With 20 sky sources in the test, there are some with unphysical coordinates which are removed in the first half of the test. I will fix the code comment, and I can also add an additional step that shows that some of the resulting `diaSource`s are unphysical if `badSourceFlags=[]`.